lib/kadm5/server_hooks.c

```diff
@@ -0,0 +1,183 @@
+    struct kadm5_hook_context **tmp;
+    kadm5_ret_t ret = KADM5_BAD_SERVER_HOOK;
+
+    hooks = krb5_config_get_strings(context, NULL,
+				    "kadmin", "hooks", NULL);
+    if (hooks == NULL)
+	return 0;
+
+    for (i = 0; hooks[i] != NULL; i++) {
+	const char *hookpath = hooks[i];
+	kadm5_hook_init_t hook_init;
+	const struct kadm5_hook *hook = NULL;
+	void *data = NULL;
+
+	handle = dlopen(hookpath, RTLD_NOW | RTLD_LOCAL | RTLD_GROUP);
+	if (handle == NULL) {
+	    krb5_warnx(context, "failed to open `%s': %s", hookpath, dlerror());
+	    ret = KADM5_SERVER_HOOK_NOT_FOUND;
+	    goto fail;
+	}
+
+	hook_init = dlsym(handle, "kadm5_hook_init");
+	if (hook_init == NULL) {
+	    krb5_warnx(context, "didn't find `kadm5_hook_init' symbol in `%s':"
+		       " %s", hookpath, dlerror());
+	    ret = KADM5_BAD_SERVER_HOOK;
+	    goto fail;
+	}
+
+	ret = hook_init(context, KADM5_HOOK_VERSION_V1, &hook, &data);
+	if (ret == 0 && hook == NULL)
+	    ret = KADM5_BAD_SERVER_HOOK;
+	if (ret) {
+	    krb5_warn(context, ret, "initialize of hook `%s' failed", hookpath);
+	    goto fail;
+	}
+
+	if (hook->version < KADM5_HOOK_VERSION_V1)
+	    ret = KADM5_OLD_SERVER_HOOK_VERSION;
+	else if (hook->version > KADM5_HOOK_VERSION_V1)
+	    ret = KADM5_NEW_SERVER_HOOK_VERSION;
+	if (ret) {
+	    krb5_warnx(context, "version of loaded hook `%s' is %u"
+		       " (supported versions %u to %u)", hookpath, hook->version,
+		       KADM5_HOOK_VERSION_V1, KADM5_HOOK_VERSION_V1);
+	    hook->fini(context, data);
+	    goto fail;
+	}
+
+	if (hook->init_context != krb5_init_context) {
+	    krb5_warnx(context, "loaded hook `%s' is not linked against "
+		       "this version of Heimdal", hookpath);
+	    hook->fini(context, data);
+	    goto fail;
+	}
+
+	hook_context = calloc(1, sizeof(*hook_context));
+	if (hook_context == NULL) {
+	    ret = krb5_enomem(context);
+	    hook->fini(context, data);
+	    goto fail;
+	}
+
+	hook_context->handle = handle;
+	hook_context->hook = hook;
+	hook_context->data = data;
+
+	tmp = realloc(ctx->hooks, (ctx->num_hooks + 1) * sizeof(*tmp));
+	if (tmp == NULL) {
+	    ret = krb5_enomem(context);
+	    hook->fini(context, data);
+	    goto fail;
+	}
+	ctx->hooks = tmp;
+	ctx->hooks[ctx->num_hooks] = hook_context;
+	hook_context = NULL;
+	ctx->num_hooks++;
+    }
+
+    return 0;
+
+fail:
+    _kadm5_s_free_hooks(ctx);
+    if (hook_context != NULL)
+	free(hook_context);
+    if (handle != NULL)
+	dlclose(handle);
+    krb5_config_free_strings(hooks);
+
+    return ret;
+}
+
+void
+_kadm5_s_free_hooks(kadm5_server_context *ctx)
+{
+    int i;
+
+    for (i = 0; i < ctx->num_hooks; i++) {
+	if (ctx->hooks[i]->hook->fini != NULL)
+	    ctx->hooks[i]->hook->fini(ctx->context, ctx->hooks[i]->data);
+	dlclose(ctx->hooks[i]->handle);
+	free(ctx->hooks[i]);
+    }
+    free(ctx->hooks);
+    ctx->hooks = NULL;
+    ctx->num_hooks = 0;
+}
+
+# else /* !HAVE_DLOPEN */
+
+kadm5_ret_t
+_kadm5_s_init_hooks(kadm5_server_context *ctx)
+{
+    return 0;
```

Comment:
In the no DLOPEN case we should log a hook load failure and return an error.

Reply:
Only if hooks are defined in krb5.conf though.